### PR TITLE
Add StringExtensions Unit Tests

### PR DESCRIPTION
**Pull request by hishamco**
In favor of #14068

### CODE REVIEW

**Comment by hishamco** [comment]
Seems there are some extensions need to be removed (unused) or replaced by native APIs

**Comment by hishamco** [comment]
Strange!! the build failed for unrelated issue

@MikeAlhayek any idea?

File: src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Utilities/StringExtensions.cs [hunk 1]
```diff
@@ -294,21 +267,16 @@ public static bool Any(this string subject, params char[] chars)
...
-        public static bool All(this string subject, params char[] chars)
+        public static bool All(this string source, params char[] chars)
         {
-            if (string.IsNullOrEmpty(subject))
-            {
-                return true;
-            }
-
-            if (chars == null || chars.Length == 0)
+            if (string.IsNullOrEmpty(source) || chars == null || chars.Length == 0)
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Utilities/StringExtensions.cs:272]: This is a wrong logic

if (chars == null || chars.Length == 0)
            {
                return false;
            }

File: src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Utilities/StringExtensions.cs [hunk 2]
```diff
@@ -11,10 +11,28 @@ namespace OrchardCore.ContentManagement.Utilities
...
+            nameof(ContentItem.Owner),
+            nameof(ContentItem.Author),
+            nameof(ContentItem.DisplayText),
+        };
+
         public static string CamelFriendly(this string camel)
         {
             // optimize common cases
-            if (string.IsNullOrWhiteSpace(camel))
+            if (String.IsNullOrWhiteSpace(camel))
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Utilities/StringExtensions.cs:35]: please use language keywords not framework. `string.` not `String.`

File: src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Utilities/StringExtensions.cs [hunk 3]
```diff
@@ -23,7 +41,7 @@ public static string CamelFriendly(this string camel)
             for (var i = 0; i < camel.Length; ++i)
             {
                 var c = camel[i];
-                if (i != 0 && char.IsUpper(c))
+                if (i != 0 && Char.IsUpper(c))
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Utilities/StringExtensions.cs:44]: language keywords

File: src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Utilities/StringExtensions.cs [hunk 4]
```diff
@@ -66,21 +85,20 @@ public static string Ellipsize(this string text, int characterCount, string elli
...
-            return string.IsNullOrEmpty(text)
+            => string.IsNullOrEmpty(text)
                 ? defaultValue
                 : new LocalizedString(null, text);
-        }
 
         public static string RemoveTags(this string html, bool htmlDecode = false)
         {
-            if (string.IsNullOrEmpty(html))
+            if (String.IsNullOrEmpty(html))
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Utilities/StringExtensions.cs:99]: language keywords

File: src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Utilities/StringExtensions.cs [hunk 5]
```diff
@@ -66,21 +85,20 @@ public static string Ellipsize(this string text, int characterCount, string elli
...
             {
-                return string.Empty;
+                return String.Empty;
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Utilities/StringExtensions.cs:101]: language keywords

File: src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Utilities/StringExtensions.cs [hunk 6]
```diff
@@ -119,16 +137,12 @@ public static string RemoveTags(this string html, bool htmlDecode = false)
...
         public static string ReplaceNewLinesWith(this string text, string replacement)
-        {
-            return string.IsNullOrWhiteSpace(text)
-                       ? string.Empty
-                       : text
-                             .Replace("\r\n", "\r\r")
-                             .Replace("\n", string.Format(replacement, "\r\n"))
-                             .Replace("\r\r", string.Format(replacement, "\r\n"));
-        }
+            => String.IsNullOrWhiteSpace(text)
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Utilities/StringExtensions.cs:140]: language keywords

File: src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Utilities/StringExtensions.cs [hunk 7]
```diff
@@ -119,16 +137,12 @@ public static string RemoveTags(this string html, bool htmlDecode = false)
...
+                ? String.Empty
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Utilities/StringExtensions.cs:141]: language keywords

File: src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Utilities/StringExtensions.cs [hunk 8]
```diff
@@ -361,23 +329,26 @@ public static string Translate(this string subject, char[] from, char[] to)
         public static string ReplaceAll(this string original, IDictionary<string, string> replacements)
         {
             var pattern = $"{string.Join("|", replacements.Keys)}";
+
             return Regex.Replace(original, pattern, match => replacements[match.Value]);
         }
 
-        public static string TrimEnd(this string rough, string trim = "")
+        public static string TrimEnd(this string source, string trimValue = "") => source.TrimEnd(trimValue.ToCharArray());
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Utilities/StringExtensions.cs:336]: source.TrimEnd(trimValue.ToCharArray()) this is different than `rough[..^trim.Length]`

**Review by MikeAlhayek — CHANGES_REQUESTED** [review]

File: src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Utilities/StringExtensions.cs (hunk 1 above)
Comment (hishamco) [reply src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Utilities/StringExtensions.cs]: Why?!! the old logic was wrong, when `source` is `null` it should returns `false` not `true`

File: src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Utilities/StringExtensions.cs [hunk 9]
```diff
@@ -180,44 +190,7 @@ public static string ToSafeName(this string name)
...
-                return true;
-            }
-
-            return false;
-        }
-
-        /// <summary>
-        /// Whether the char is a letter between A and Z or not.
-        /// </summary>
-        public static bool IsLetter(this char c)
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Utilities/StringExtensions.cs:212]: mark these 2 methods obsolete and keep them just so we don't break anyone. you can change the logic to return `char.IsLetter(c)` and `chat.IsWhitespace()`

File: src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Utilities/StringExtensions.cs (hunk 1 above)
Comment (hishamco) [reply src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Utilities/StringExtensions.cs:272]: There was no test before, and that is why the test failed when I use the previous logic

**Comment by Piedone** [comment]
Do you want to get back to reviewing this, @MikeAlhayek?

**Comment by hishamco** [comment]
Anything to add here or shall I merge this?

**Pull request merged**